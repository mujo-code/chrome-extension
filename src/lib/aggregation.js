import { shortURL } from './url'
import { set } from './util'

const combineUrlTimes = data => {
  const keys = Object.keys(data)
  return keys.reduce((accum, key) => {
    const shortKey = shortURL(key)
    set(accum, shortKey, data[key] + (accum[shortKey] || 0))
    return accum
  }, {})
}

const getTotalTime = data => {
  const iterable = Object.keys(data)
  return iterable
    .map(key => data[key])
    .reduce((total, value) => {
      /* eslint-disable-next-line no-param-reassign */
      total += value
      return total
    }, 0)
}

export const siteTimeToChartData = rawData => {
  const data = combineUrlTimes(rawData)
  const totals = getTotalTime(data)
  return Object.keys(data).map(label => {
    const percent = data[label] / totals
    return {
      label,
      percent,
      originalData: { [label]: data[label] },
    }
  })
}
