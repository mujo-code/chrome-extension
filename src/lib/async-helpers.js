export const wait = (amount = 0) =>
  new Promise(resolve => {
    setTimeout(resolve, amount)
  })

export const defer = async (fn, amount, ...args) => {
  await wait(amount)
  return fn(...args)
}

export const composePromises = (...fns) => value =>
  fns.reduceRight(
    (promise, fn) => promise.then(resolve => fn(resolve)),
    Promise.resolve(value)
  )

export const pipePromises = (...fns) => composePromises(fns.reverse())
