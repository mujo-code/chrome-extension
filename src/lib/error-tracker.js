import * as Sentry from '@sentry/browser'

const { SENTRY_DSN } = process.env

export const createTracker = ({ dsn, environment }) => {
  const hasDSN = !!dsn
  if (hasDSN) {
    Sentry.init({ dsn, environment })
  }

  return {
    exception(error) {
      if (hasDSN) return null
      return Sentry.captureException(error)
    },
    async exceptionWithInfo(error, errorInfo) {
      if (hasDSN) return null
      return new Promise(resolve => {
        Sentry.withScope(scope => {
          scope.setExtras(errorInfo)
          return resolve(Sentry.captureException(error))
        })
      })
    },
  }
}

export const tracker = createTracker({ dsn: SENTRY_DSN })
