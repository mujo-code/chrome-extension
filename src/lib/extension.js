import { GET_STORAGE, SET_STORAGE } from '../constants'
import { promisifyObject } from './promisify'
/*
  Extension Lib
  -----
  Essentially a way to abstract away chrome bits
  to eventually allow for bits like firefox without large refactors
*/

export const message = async (event, data) =>
  new Promise((resolve, reject) => {
    const payload = Object.assign({ event }, data || {})
    chrome.runtime.sendMessage(payload, (response = {}, ...args) => {
      console.log([response], args)
      if (response.error) {
        reject(response.error)
        return
      }
      resolve(response)
    })
  })

export const getStorage = async key => {
  const response = await message(GET_STORAGE, { key })
  return response.value
}

export const setStorage = async (key, value) => {
  const response = await message(SET_STORAGE, { key, value })
  return response.value
}

export const onMessage = (...args) => {
  chrome.runtime.onMessage.addListener(...args)
}

export const permissions = promisifyObject(chrome.permissions)

export const {
  alarms,
  notifications,
  tabs,
  webNavigation,
  runtime,
  topSites,
} = chrome
