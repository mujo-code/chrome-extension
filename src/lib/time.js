// format MS to M, S, H
export const factor = (units, conv) => Math.floor(units / conv)

export const getTime = ms => {
  const seconds = factor(ms, 1000)
  const minutes = factor(seconds, 60)
  const hours = factor(minutes, 60)
  return {
    hours,
    minutes: minutes - hours * 60,
    seconds: seconds - minutes * 60,
    ms: ms - seconds * 1000,
  }
}

export const minutesToMS = minutes => minutes * 60 * 1000

export const readableTime = num => {
  if (typeof num !== 'number') {
    throw new Error(
      `Children must be a "number" in <Time /> Component got: ${typeof num}`
    )
  }
  const time = getTime(num)
  const timeStr = [] // but array ðŸ¤”
  if (time.hours) {
    timeStr.push(`${time.hours} hour${time.hours > 1 ? 's' : ''}`)
  }
  if (time.minutes) {
    timeStr.push(
      `${time.minutes} minute${time.minutes > 1 ? 's' : ''}`
    )
  }
  if (time.seconds) {
    timeStr.push(
      `${time.seconds} second${time.seconds > 1 ? 's' : ''}`
    )
  }
  if (timeStr.length === 0) {
    return `${Math.floor(time.ms)} milliseconds`
  }
  return [timeStr.slice(0, -1).join(', '), timeStr.pop()]
    .filter(x => x)
    .join(' and ')
}
