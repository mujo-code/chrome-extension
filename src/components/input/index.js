import { Box } from '@mujo/box'
import { removeKeys } from '@mujo/box/dist/lib/remove-keys'
import React from 'react'
import { useTheme } from '../../hooks/use-theme'
import { toDashCase } from '../../lib/util'
import { HeaderS } from '../fonts'

export const Input = props => {
  const { foreground, background } = useTheme()
  const { label, id: propsId, marginTop } = props
  const id = propsId || toDashCase(label)
  const otherProps = removeKeys(props, 'label', 'id', 'marginTop')
  return (
    <Box display="flex" direction="column">
      <HeaderS
        Component="label"
        htmlFor={id}
        color={foreground}
        marginBottom="zero"
        marginTop={marginTop}
      >
        <Box Component="small">{label}</Box>
      </HeaderS>
      <Box
        color={props.color || background}
        backgroundColor={props.backgroundColor || foreground}
        border="none"
        marginTop="xs"
        id={id}
        {...otherProps}
      />
    </Box>
  )
}

Input.defaultProps = {
  Component: 'input',
  type: 'text',
  paddingLeft: 'm',
  paddingRight: 'm',
  paddingTop: 'xs',
  paddingBottom: 'xs',
  marginTop: 'm',
  marginBottom: 'm',
  borderRadius: 's',
  label: 'Label is Required',
}
