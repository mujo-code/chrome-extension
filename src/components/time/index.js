import { Box } from '@jcblw/box'
import { removeKeys } from '@jcblw/box/dist/lib/remove-keys'
import React from 'react'
import { readableTime } from '../../lib/time'

export const Time = props => (
  <Box {...removeKeys(props, 'children')}>
    {readableTime(props.children)}
  </Box>
)
