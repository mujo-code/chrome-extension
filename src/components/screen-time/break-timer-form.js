import { Box } from '@mujo/box'
import React from 'react'
import { FOURTY_FIVE_MINUTES } from '../../constants'
import { msToMinutes, minutesToMS } from '../../lib/time'
import { HeaderS } from '../fonts'
import { Input } from '../input'
import { Switch } from '../switch'

export const BreakTimerForm = props => {
  const { time, enabled, originalURL, setBreakTimer } = props
  const minutes = msToMinutes(time)
  const suffix = minutes === 1 ? '' : 's'
  return (
    <Box display="flex" direction="column">
      {enabled ? (
        <Box
          display="flex"
          direction="column"
          marginRight="s"
          alignItems="flexStart"
        >
          <Input
            marginTop="zero"
            marginBottom="m"
            id="break-timer-time"
            label={`Shown in ${minutes} minute${suffix} of usage`}
            type="number"
            value={minutes}
            onChange={e =>
              setBreakTimer(
                originalURL,
                minutesToMS(
                  Math.abs(parseInt(e.target.value || '0', 10))
                ),
                enabled
              )
            }
          />
        </Box>
      ) : null}
      <Box
        display="flex"
        direction="row"
        alignItems="center"
        marginRight="s"
        flex="1"
      >
        <Switch
          id="break-timer"
          marginRight="s"
          value={enabled}
          onChange={() =>
            setBreakTimer(
              originalURL,
              time || FOURTY_FIVE_MINUTES,
              !enabled
            )
          }
        />
        <HeaderS
          marginTop="zero"
          marginBottom="zero"
          labelFor="break-timer"
        >
          Break Timer
        </HeaderS>
      </Box>
    </Box>
  )
}
