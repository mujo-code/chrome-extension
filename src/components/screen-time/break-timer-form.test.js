import React from 'react'
import { create } from 'react-test-renderer'
import { minutesToMS } from '../../lib/time'
import { Input } from '../input'
import { BreakTimerForm } from './break-timer-form'

test('BreakTimerForm should match snapshot', () => {
  const tree = create(<BreakTimerForm />).toJSON()
  expect(tree).toMatchSnapshot()
})

test('BreakTimerForm should match snapshot when enabled', () => {
  const tree = create(<BreakTimerForm enabled={true} />).toJSON()
  expect(tree).toMatchSnapshot()
})

test('BreakerTimerForm input should be positive', () => {
  const fn = jest.fn()
  const originalURL = 'foo'
  const enabled = true
  const wrapper = create(
    <BreakTimerForm
      originalURL={originalURL}
      setBreakTimer={fn}
      enabled={enabled}
    />
  )
  const { root } = wrapper
  const input = root.findByType(Input)

  input.props.onChange({ target: { value: -11 } })

  expect(fn).toBeCalledWith(originalURL, minutesToMS(11), enabled)
})
