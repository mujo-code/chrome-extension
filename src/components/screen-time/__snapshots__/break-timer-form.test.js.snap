// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreakTimerForm should match snapshot 1`] = `
<div
  className="css-j7qwjs"
>
  <div
    className="css-102g4jn"
  >
    <input
      className="css-1ay9vb9"
      id="break-timer"
      onChange={[Function]}
      type="checkbox"
    />
    <h4
      className="css-11ejjzl"
      color="color"
      labelFor="break-timer"
    >
      Break Timer
    </h4>
  </div>
</div>
`;

exports[`BreakTimerForm should match snapshot when enabled 1`] = `
<div
  className="css-j7qwjs"
>
  <div
    className="css-102g4jn"
  >
    <input
      checked={true}
      className="css-1ay9vb9"
      id="break-timer"
      onChange={[Function]}
      type="checkbox"
    />
    <h4
      className="css-11ejjzl"
      color="color"
      labelFor="break-timer"
    >
      Break Timer
    </h4>
  </div>
  <div
    className="css-1lqrh8c"
  >
    <div
      className="css-j7qwjs"
    >
      <label
        className="css-11ejjzl"
        color="mischka"
        htmlFor="break-timer-time"
      >
        <small
          className="css-0"
        >
          Shown in NaN minutes of usage
        </small>
      </label>
      <input
        backgroundColor="mischka"
        className="css-1hi28e2"
        color="outerSpace"
        id="break-timer-time"
        onChange={[Function]}
        type="number"
        value={NaN}
      />
    </div>
  </div>
</div>
`;
